new/rfcXXX-policyfile-endpoint.md

```diff
@@ -0,0 +1,482 @@
+converge a node (host) to a desired state. In contrast with the current
+behaviors of Chef Client and Server, Policyfiles provide a way to
+describe and manage configuration code on a whole-node basis, reducing
+Chef Client's use of shared, mutable resources hosted on the server and
+providing operators easier visibility into configuration code and
+changes. For further description of the Policyfile feature, read the dedicated
+[POLICYFILE README.](https://github.com/chef/chef-dk/blob/master/POLICYFILE_README.md)
+
+A preview implementation of the Policyfile feature is currently available; in
+this implementation, Policyfile documents are stored as data bag items.
+Although this implementation works, it does not provide the baseline quality
+users expect. In particular, there is no validation, no means to make any
+relational queries, and documents are indexed for search as data bag items.
+
+This RFC describes a new Policyfile HTTP Resource API that remediates these
+issues.
+
+## Motivation
+
+    As a Chef user,
+    I want Policyfiles to have a dedicated HTTP Resource API,
+    so that Policyfiles will be validated upon upload
+    and I can query the relationships between Policyfiles and other objects,
+    and Policyfiles will not be indexed as an incorrect data type.
+
+## Specification
+
+### General Concepts
+
+A Policyfile lock is a JSON document that specifies a node's run list and the
+set of cookbooks a node will use to converge the node.
+
+A Policy Group is a set of nodes defined by a single token. Nodes must be a
+member of a Policy Group in order to use the Policyfile feature, and nodes may
+only be a member of one policy group.
+
+### Policyfile Lock
+
+A Policyfile lock is a JSON document that has the following structure:
+
+```json
+{
+  "revision_id": "edd40c30c4e0ebb3658abde4620597597d2e9c17",
+  "name": "some_policy_name",
+  "run_list": [
+    "recipe[policyfile_example_cookbook::default]"
+  ],
+  "cookbook_locks": {
+    "policyfile_example_cookbook": {
+      "version": "1.0.0",
+      "identifier": "f04cc40faf628253fe7d9566d66a1733fb1afbe9"
+    }
+  }
+}
+
+```
+
+#### `revision_id` Field (required)
+
+The `revision_id` field is used to distinguish revisions of a Policyfile lock
+document. It is a JSON String. It may contain alphanumeric characters, hyphens,
+underscores, the dot character and the colon character. In regular expression
+form: `/^[\-[:alnum:]_\.\:]+$/`. The revision ID is limited to a maximum of 255
+characters. The Chef Server must reject a Policyfile lock document when the
+revision ID does not meet these requirements.
+
+In ChefDK the revision will be calculated by hashing other content in the
+Policyfile lock document and thus will be a hexidecimal number. In order to
+allow alternate implementations to generate revision IDs according to different
+schemes (perhaps timestamps or Ci build numbers), this field will not be
+required to be a hexidecimal string.
```

Comment:
@manderson26 I'm coming around to thinking that what you're requesting should at least be possible.

My concern is that the current implementation isn't designed to provide strong integrity guarantees. For example, cookbook identifiers are a SHA-1 of the MD5s of the files in the cookbook (code here: https://github.com/chef/chef-dk/blob/b227c377703625ba5da2744d5dbc623c6ba923be/lib/chef-dk/cookbook_profiler/identifiers.rb#L41-L64) which is fine for well-intentioned users, but I would certainly change to a SHA-2 of SHA-2s if I were to make any claims about integrity. Likewise, I would make the revision id a SHA-2 of the canonicalized content if I wanted to make claims about integrity.

Speaking of the canonicalized content, the current implementation of this doesn't account for attributes at all (code: https://github.com/chef/chef-dk/blob/b227c377703625ba5da2744d5dbc623c6ba923be/lib/chef-dk/policyfile_lock.rb#L145-L176), which it certainly will have to when that feature is added. Also I could be doing something dumb in the canonicalization process.

Given that I'd like to be shipping sooner rather than later, that means I can either spec out two behaviors or just ship the feature in violation of the spec and get into compliance later.

Another thought I had (and maybe this is terrible) is that API consumers opt-in to strict crypto checking by adding a `!` to the end of the cookbook identifier or policyfile lock revision ID. Since this character is currently reserved, this would be backwards (but not forwards) compatible, and since the character is part of the ID, it would prevent the delete/recreate threat you described.